### PR TITLE
[kong] add status listen

### PR DESCRIPTION
**Pull request by rainest**
#### What this PR does / why we need it:
Add the status listen, which provides a dedicated status endpoint on versions 1.4+.

#### Special notes for your reviewer:
kong.conf does not provide a TLS listen in the example. The chart adds one with a somewhat arbitrary port choice.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] PR is based off the current tip of the `next` branch and targets `next`, not `master`
- [x] New or modified sections of values.yaml are documented in the README.md
- [x] Title of the PR and commit headers start with chart name (e.g. `[kong]`)

### CODE REVIEW

**Comment by rainest** [comment]
`ssl` support is lacking certificate directives in nginx_kong.lua. We should consider whether to fix that first.

**Comment by rainest** [comment]
Force-pushed to reconcile conflicts with next after merging stream listen.

Having thought over this a bit more, I think we should default HTTPS to false if we want to release this for 2.0.x and adhere to what we claim in the appVersion. Even a future minor release adds HTTPS support, I'm wary of introducing changes that break startup on releases that do fall under what the chart indicates support for. If we want to default to true, we should move towards supporting that in 2.1 and defer this until then.

Added a core backlog item to discuss that.

File: charts/kong-collectorapi/glum-lizard.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+image:
+  repository: gcr.io/formidable-app-197923/kong-brain-immunity
```
Comment (hbagdi) [inline charts/kong-collectorapi/glum-lizard.yaml]: Please remove these collector related files.

File: charts/kong/README.md [hunk 2]
```diff
@@ -300,6 +300,11 @@ individual services: see values.yaml for their individual default values.
 * `manager`
 * `portal`
 * `portalapi`
+* `status`
+
+Unlike most services, `status` cannot be exposed outside the pod, and cannot
+create a Kubernetes service or ingress. It supports the settings under
+`SVC.http` and `SVC.tls` only.
```
Comment (hbagdi) [inline charts/kong/README.md]: > Unlike most services

Unlike all other services

> cannot be exposed outside the pod, and cannot
create a Kubernetes service or ingress

Can we instead frame this as "The chart doesn't allow exposing the endpoint as outside the Kubernetes cluster because the endpoint is meant for intra-cluster concerns only.".

File: charts/questions.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+* Would there be any reason not to generate a service for each listen? Bit cluttered, but whatever.
+* Generic listens present a challenge: there is a static list of possible X_listen parameters, and shouldn't really be user-configurable. We don't have any way to prevent users from just removing or changing them. Probably better to keep dedicated blocks and maybe reuse templates underneath.
```
Comment (hbagdi) [inline charts/questions.txt]: Please remove this file.

File: charts/kong-collectorapi/collector-notes.txt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# Non-agnostic README/NOTES items
+* `open` is OS X-only.
+* Should not assume Minikube.
+* Should not assume open NodePort. We're guilty of this too--we should probably add a port-forward.
```
Comment (hbagdi) [inline charts/kong-collectorapi/collector-notes.txt]: Please remove this file.

File: charts/kong/templates/_helpers.tpl [hunk 5]
```diff
@@ -109,7 +109,7 @@ Create a single listen (IP+port+parameter combo)
 {{- define "kong.singleListen" -}}
   {{- $listen := list -}}
   {{- $listen = append $listen (printf "0.0.0.0:%d" (int64 .containerPort)) -}}
-  {{- range $param := .parameters | uniq }}
+  {{- range $param := .parameters | default (list) | uniq }}
```
Comment (hbagdi) [inline charts/kong/templates/_helpers.tpl:112]: What is the reason for this change?
Does this error out if parameters is undefined?

File: charts/kong/values.yaml [hunk 6]
```diff
@@ -101,6 +101,20 @@ admin:
...
+  http:
+    # Enable plaintext HTTP listen for the status listen
+    enabled: true
+    containerPort: 8100
+
+  tls:
+    # Enable HTTPS listen for the status listen
+    # Kong does not currently support HTTPS status listens, so this should remain false
+    enabled: false
+    containerPort: 8143
```
Comment (hbagdi) [inline charts/kong/values.yaml]: nit: Since Kong adds 443 to a port for TLS, let's change this to 8543?

**Review by hbagdi — CHANGES_REQUESTED** [review]

File: charts/kong/templates/_helpers.tpl (hunk 5 above)
Comment (rainest) [reply charts/kong/templates/_helpers.tpl:112]: Yeah, see the last commit/CI log for the commit prior. Range over null generates an error.

File: charts/kong/README.md (hunk 2 above)
Comment (rainest) [reply charts/kong/README.md]: There's no service creation either--it's not accessible outside the pod, barring things like readiness/liveness checks. Added the why in front, but I think we should keep the specifics of what it does--K8S-saavy users should recognize what no service/no ingress means in practice.

File: charts/kong/README.md [hunk 7]
```diff
@@ -300,6 +300,12 @@ individual services: see values.yaml for their individual default values.
 * `manager`
 * `portal`
 * `portalapi`
+* `status`
+
+`status` is intended for internal use within the Kong pod only. Unlike other
```
Comment (hbagdi) [inline charts/kong/README.md]: > internal use within the Kong pod only

I think you meant within the internal Kubernetes cluster only.

Comment (hbagdi) [reply charts/kong/README.md]: status exposes health and metrics endpoints that are for in-cluster services only.

**Review by hbagdi — APPROVED** [review]
